File: 20-valid-parentheses/20-valid-parentheses.cs
public class Solution {
    public bool IsValid(string s) {
        
          Dictionary<char,char> map = new Dictionary<char,char>();
          map.Add(')','(');
          map.Add('}','{');
          map.Add(']','[');
        
          Stack<char> stack = new Stack<char>();
        
          for(int i=0;i < s.Length; i++)
          {
              if(map.ContainsKey(s[i]))
              {
                  char top = stack.Count == 0 ? '#' : stack.Pop();
                  
                  if(top != map[s[i]])
                      return false;
              }
              else
              {
                  stack.Push(s[i]);
              }
              
          }
        
          return !stack.Any();
        
    }
}

File: 20.valid-parentheses.java
import java.util.HashMap;
import java.util.Stack;

/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {

    private HashMap<Character,Character> map; 

    public Solution()
    {
        map = new HashMap<>();

        map.put(')','(');
        map.put('}','{');
        map.put(']','[');
    }

    public boolean isValid(String s) {
     
        Stack<Character> stack = new Stack<Character>();

        for(int i=0;i < s.length();i++)
        {
             char c = s.charAt(i);

            if(map.containsKey(c))
            {
                char topElement = stack.empty() ? '#' : stack.pop();

                if(topElement != map.get(c))
                   return false;

            }
            else
            {
                stack.push(c);
            }

        }
        
        return stack.isEmpty();

    }
}
// @lc code=end

